packages/flutter/lib/src/widgets/sliver_coordinator.dart

```diff
@@ -0,0 +1,254 @@
+/// scroll notifications and the concomittant layout changes in the scroll view's
+/// [CoordinatedSliver]s. It cannot be used to change sliver layouts. Typical use cases:
+///
+/// - When a drag gesture ends: auto-scrolling a widget that overlaps the top
+/// or bottom of the viewport so that it's completely visible.
+/// - When a sliver scrolls under a pinned header, change the header's
+/// content or elevation with an animated transition.
+/// - Show an indicator while a special sliver is visible or when its
+/// visibility changes.
+///
+/// {@tool dartpad}
+/// This example contains one [CoordinatedSliver] which is
+/// auto-scrolled so that it's aligned with the top of the viewport
+/// whenever a scroll-end gesture leaves it partially visible. The
+/// auto-scroll is triggered by the [SliverCoordinator]'s
+/// callback. The callback has access to the current scroll offset and
+/// the current extent (height) of the [CoordinatedSliver].
+///
+/// ** See code in examples/api/lib/widgets/sliver/sliver_coordinator.0.dart **
+/// {@end-tool}
+class SliverCoordinator extends StatefulWidget {
+  /// Create specialized scroll notification listener for a [CustomScrollView] descendant.
+  ///
+  /// The [callback] runs each time the scroll view is scrolled.
+  const SliverCoordinator({ super.key, required this.callback, required this.child });
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+
+  /// Called each time the [CustomScrollView] below this widget scrolls.
+  ///
+  /// This callback runs after its descendants, including the scroll
+  /// view, have been laid out.  The callback's
+  /// [SliverCoordinatorData] parameter includes the
+  /// [SliverConstraints] and [SliverGeometry] computed for each
+  /// [CoordinatedSliver].
+  final SliverCoordinatorCallback callback;
+
+  @override
+  State<SliverCoordinator> createState() => _SliverCoordinatorState();
+
+  static SliverCoordinatorData _of(BuildContext context) {
+    final _SliverCoordinatorScope? scope = context.dependOnInheritedWidgetOfExactType<_SliverCoordinatorScope>();
+    return scope!.data;
+  }
+}
+
+class _SliverCoordinatorState extends State<SliverCoordinator> {
+  SliverCoordinatorData data = SliverCoordinatorData();
+
+  bool handleScrollNotification(ScrollNotification notification) {
+    // The table of CoordinatedSliver data will be repopulated after the
+    // scroll view is laid out (see RenderCoordinatedSliver.performLayout)
+    // and before the callback runs.
+    if (notification is ScrollUpdateNotification) {
+      data._clear();
+    }
+    // The callback runs after the descendant CustomScrollView's viewport
+    // has been laid out and the _SliverLayoutInfo object has been updated.
+    SchedulerBinding.instance.addPostFrameCallback((Duration duration) {
+      widget.callback(notification, data);
+    });
```

Comment:
We should document this frame delay, explaining we need to wait until after layout.

Reply:
Oh! We should probably check we're still mounted here. That gets us in these post frame callbacks. :)

Reply:
YES